### PR TITLE
Added LRU cache, web worker, and isolated encryption algorithm logic …

### PR DESCRIPTION
**Pull request by Raunak-Agrawal**

### CODE REVIEW

File: constants/narcissistNumbers.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
+const NARCISSIT_NUMBERS = {
```
Comment (ankushdharkar) [inline constants/narcissistNumbers.js]: Please remove this file

File: helpers/index.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export const joinArrayAsString = (arr) => {
```
Comment (ankushdharkar) [inline helpers/index.js]: Why helper? We already have this method in Array prototype. Please remove

File: index.html [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<meta charset="UTF-8" />
+		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
+		<title>Secret Scientist</title>
+		<!-- <link rel="stylesheet" type="text/css" href="style.css" /> -->
```
Comment (ankushdharkar) [inline index.html]: Please remove dead code

File: modules/decryption.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+			resultArray.push(stringPlusOne);
+		}
+		if (numPlusOne > MAX_TOKEN || i === string.length - 1) {
+			resultArray.push([string[i]]);
+			i = i + 1;
+		} else {
+			const numString = string[i] + string[i + 1];
+			resultArray.push([numString]);
+			i = i + 2;
+		}
```
Comment (ankushdharkar) [inline modules/decryption.js]: Please break into smaller functions. Logic is becoming convoluted

File: modules/decryption.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+function mapEncryptedLetterToNumbers(letters) {
+	const numberArray = [];
+	for (let letter of letters) {
+		numberArray.push(letter.toLowerCase().charCodeAt(0) - 97);
+	}
+	return joinArrayAsString(numberArray);
+}
+
+function decryptionModule(inputValue, key) {
+	KEY = key;
```
Comment (ankushdharkar) [inline modules/decryption.js]: Confusing assignment to global. Please avoid

File: modules/encryption.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ * @param {array} value -  Mapped array of numbers with key; [10,11,12]
+ * @return {array} - Mapped array of encrypted numbers; [153,370,371]
+ */
+
+function encryptNumbersWithAlgorithm(numbersWithKey) {
+	return new Promise((resolve, reject) => {
+		const myWorker = new Worker();
+		myWorker.postMessage(numbersWithKey);
+		myWorker.addEventListener('message', (event) => resolve(event.data));
+		myWorker.addEventListener('error', reject);
```
Comment (ankushdharkar) [inline modules/encryption.js]: Separate out messaging API from Business Logic please

File: modules/encryption.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+function encyptionModule(word) {
+	const numberArray = mapWordToNumbers(word);
+	const numbersWithKey = addNumbersWithKey(numberArray, KEY);
+	console.log(numbersWithKey, 'numbersWithKey');
+	return new Promise((resolve, reject) => {
+		encryptNumbersWithAlgorithm(numbersWithKey)
+			.then((res) => {
+				console.log(res, 'RES');
```
Comment (ankushdharkar) [inline modules/encryption.js]: Remove please

File: script.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import encyptionModule from './modules/encryption';
+import decryptionModule from './modules/decryption';
+import './style.css';
+
+function onClick(operationType) {
+	const inputVal = document.querySelector('.inputstring').value;
+	if (operationType === 'encrypt') {
+		if (!/^[a-zA-Z]+$/.test(inputVal)) {
+			alert('Please input just characters');
+			return;
```
Comment (ankushdharkar) [inline script.js]: Cool handling. Move to separate function please

File: webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+					options: {
+						presets: ['@babel/preset-env'],
+					},
+				},
+			},
+			{
+				test: /\.worker\.js$/,
+				exclude: /(node_modules)/,
+				use: {
+					loader: 'worker-loader',
```
Comment (ankushdharkar) [inline webpack.config.js:34]: Awesome

File: webworker/workers.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+	n ? digit_pow(Math.floor(n / 10), pow, sum + (n % 10) ** pow) : sum;
+
+function nnGenerator(num) {
+	const numberArray = new Array(num);
+	let count = 0;
+	let i = 1;
+	while (i) {
+		if (digit_pow(i) === i) {
+			if (numberArray[num - 1]) {
+				break;
```
Comment (ankushdharkar) [inline webworker/workers.js]: Avoid please. Causes confusion in code. Reduces predictable readability

**Review by ankushdharkar — CHANGES_REQUESTED** [review]
Please make the changes are re-request a review

File: constants/narcissistNumbers.js (hunk 1 above)
Comment (Raunak-Agrawal) [reply constants/narcissistNumbers.js]: ok

File: helpers/index.js (hunk 2 above)
Comment (Raunak-Agrawal) [reply helpers/index.js]: Ok

File: index.html (hunk 3 above)
Comment (Raunak-Agrawal) [reply index.html]: Sure

File: modules/decryption.js (hunk 4 above)
Comment (Raunak-Agrawal) [reply modules/decryption.js]: Removing the file itself. It is a decryption which I believe we are not doing at this point

File: modules/decryption.js (hunk 5 above)
Comment (Raunak-Agrawal) [reply modules/decryption.js]: Removed

File: modules/encryption.js (hunk 7 above)
Comment (Raunak-Agrawal) [reply modules/encryption.js]: Sure

File: script.js (hunk 8 above)
Comment (Raunak-Agrawal) [reply script.js]: ok

File: webpack.config.js (hunk 9 above)
Comment (Raunak-Agrawal) [reply webpack.config.js:34]: Thanks :)

File: webworker/workers.js (hunk 10 above)
Comment (Raunak-Agrawal) [reply webworker/workers.js]: Ok

**Pull request closed without merging**